### ISSUE TITLE
Eval bug: llama-perplexity segfaults and fails to produce logits at long-context (20k context)

### ISSUE THREAD

**Issue by compilebunny** [ggml-org/llama.cpp#23569]
### Name and Version

llama-perplexity --version
version: 9295 (95405ac65)
built with GNU 15.2.0 for Linux x86_64

### Operating systems

Linux

### GGML backends

CPU

### Hardware

First failed with CUDA using "--fit on", then failed with CUDA using fixed number of layers on GPU. Finally, tried with CPU-only (128 gb RAM) to eliminate the possibility of memory limitations. Failed there too. The logs provided are for CPU only.

### Models

Qwen3.5-9B-BF16.gguf

### Problem description & steps to reproduce

I am attempting to use llama-perplexity to evaluate model performance at long contexts. 

The following command:

`CUDA_VISIBLE_DEVICES="" llama-perplexity -m Qwen3.5-9B-BF16.gguf -f ./logits_temp/txt_ctx20480_ch0.txt -c 20480 -t 10 --fit off -ngl 0 -b 1024 --no-warmup -ctk bf16 -ctv bf16 --kl-divergence-base ./logits_temp/gt_Qwen3.5-9B-BF16.gguf_ctx20480_ch0.bin -v
`

produces a logits file that is a few hundred KB in length and exits with code 139. The logits file for the same command with a context of 10240 produces correctly @ 10 GB, so I would expect this file to be 20 GB.

```
0.00.654.976 D llama_context: enumerating backends
0.00.654.983 D llama_context: backend_ptrs.size() = 1
0.00.654.987 I sched_reserve: reserving ...
0.00.654.995 D sched_reserve: max_nodes = 20480
0.00.660.255 D sched_reserve: reserving full memory module
0.00.660.270 D sched_reserve: worst-case: n_tokens = 512, n_seqs = 1, n_outputs = 1
0.00.660.274 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.00.660.700 I sched_reserve: Flash Attention was auto, set to enabled
0.00.660.705 I sched_reserve: resolving fused Gated Delta Net support:
0.00.660.709 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.00.661.024 I sched_reserve: fused Gated Delta Net (autoregressive) enabled
0.00.661.030 D graph_reserve: reserving a graph for ubatch with n_tokens =   16, n_seqs =  1, n_outputs =   16
0.00.661.328 I sched_reserve: fused Gated Delta Net (chunked) enabled
0.00.661.334 D graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =  512
0.00.661.893 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.00.662.352 D graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =  512
0.00.662.836 I sched_reserve:        CPU compute buffer size =   509.00 MiB
0.00.662.842 I sched_reserve: graph nodes  = 1856
0.00.662.845 I sched_reserve: graph splits = 1
0.00.662.850 I sched_reserve: reserve took 7.86 ms, sched copies = 1
0.00.662.929 D set_adapters_lora: adapters = (nil)
0.00.662.935 D adapters_lora_are_same: adapters = (nil)
0.00.662.941 I 
0.00.663.012 I system_info: n_threads = 10 (n_threads_batch = 10) / 16 | CUDA : ARCHS = 900 | USE_GRAPHS = 1 | PEER_MAX_BATCH_SIZE = 128 | FA_ALL_QUANTS = 1 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX_VNNI = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 | 
0.00.663.059 I perplexity: saving all logits to ./logits_temp/gt_Qwen3.5-9B-BF16.gguf_ctx20480_ch0.bin
0.00.663.069 I perplexity: tokenizing the input ..
0.00.792.515 I perplexity: tokenization took 129.44 ms
0.00.792.593 I perplexity: calculating perplexity over 4 chunks, n_ctx=20480, batch_size=1024, n_seq=1
18.56.929.947 I perplexity: 1133.08 seconds per pass - ETA 1 hours 15.53 minutes
Segmentation fault (core dumped)
```

### Relevant log output

[large-logit-attempt.txt](https://github.com/user-attachments/files/28175165/large-logit-attempt.txt)

<details>
<summary>Logs</summary>

```console
0.00.654.976 D llama_context: enumerating backends
0.00.654.983 D llama_context: backend_ptrs.size() = 1
0.00.654.987 I sched_reserve: reserving ...
0.00.654.995 D sched_reserve: max_nodes = 20480
0.00.660.255 D sched_reserve: reserving full memory module
0.00.660.270 D sched_reserve: worst-case: n_tokens = 512, n_seqs = 1, n_outputs = 1
0.00.660.274 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.00.660.700 I sched_reserve: Flash Attention was auto, set to enabled
0.00.660.705 I sched_reserve: resolving fused Gated Delta Net support:
0.00.660.709 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.00.661.024 I sched_reserve: fused Gated Delta Net (autoregressive) enabled
0.00.661.030 D graph_reserve: reserving a graph for ubatch with n_tokens =   16, n_seqs =  1, n_outputs =   16
0.00.661.328 I sched_reserve: fused Gated Delta Net (chunked) enabled
0.00.661.334 D graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =  512
0.00.661.893 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.00.662.352 D graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =  512
0.00.662.836 I sched_reserve:        CPU compute buffer size =   509.00 MiB
0.00.662.842 I sched_reserve: graph nodes  = 1856
0.00.662.845 I sched_reserve: graph splits = 1
0.00.662.850 I sched_reserve: reserve took 7.86 ms, sched copies = 1
0.00.662.929 D set_adapters_lora: adapters = (nil)
0.00.662.935 D adapters_lora_are_same: adapters = (nil)
0.00.662.941 I 
0.00.663.012 I system_info: n_threads = 10 (n_threads_batch = 10) / 16 | CUDA : ARCHS = 900 | USE_GRAPHS = 1 | PEER_MAX_BATCH_SIZE = 128 | FA_ALL_QUANTS = 1 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX_VNNI = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 | 
0.00.663.059 I perplexity: saving all logits to ./logits_temp/gt_Qwen3.5-9B-BF16.gguf_ctx20480_ch0.bin
0.00.663.069 I perplexity: tokenizing the input ..
0.00.792.515 I perplexity: tokenization took 129.44 ms
0.00.792.593 I perplexity: calculating perplexity over 4 chunks, n_ctx=20480, batch_size=1024, n_seq=1
18.56.929.947 I perplexity: 1133.08 seconds per pass - ETA 1 hours 15.53 minutes
Segmentation fault (core dumped)
```
</details>

**Comment by feffy380** [comment]
Encountered the same segfault on b9292 (ef570f6) with Qwen3.6 27B at 32k context, which leads me to believe the length overflow bug isn't completely fixed. 16k context works fine

**Comment by fairydreaming** [comment]
@compilebunny @feffy380 

You are right, it looks like there are more integer overflows to fix. In the meantime you can try this patch:

```
diff --git a/tools/perplexity/perplexity.cpp b/tools/perplexity/perplexity.cpp
index ea9de83e1..027cd53f2 100644
--- a/tools/perplexity/perplexity.cpp
+++ b/tools/perplexity/perplexity.cpp
@@ -157,7 +157,7 @@ static void process_logits(std::ostream& out, int n_vocab, const float * logits,
                 break;
             }
             lock.unlock();
-            const double v = log_softmax(n_vocab, logits + size_t(i)*n_vocab, log_probs.data() + i*nv, tokens[i+1]);
+            const double v = log_softmax(n_vocab, logits + size_t(i)*n_vocab, log_probs.data() + size_t(i)*nv, tokens[i+1]);
             local_nll += v;
             local_nll2 += v*v;
         }
@@ -169,7 +169,7 @@ static void process_logits(std::ostream& out, int n_vocab, const float * logits,
     for (auto & w : workers) {
         w.join();
     }
-    out.write((const char *)log_probs.data(), n_token*nv*sizeof(uint16_t));
+    out.write((const char *)log_probs.data(), size_t(n_token)*nv*sizeof(uint16_t));
 }
 
 struct kl_divergence_result {
@@ -279,7 +279,7 @@ static void process_logits(int n_vocab, const float * logits, const int * tokens
                 break;
             }
             lock.unlock();
-            std::pair<double, float> v = log_softmax(n_vocab, logits + size_t(i)*n_vocab, base_log_probs.data() + i*nv, tokens[i+1], local_kld);
+            std::pair<double, float> v = log_softmax(n_vocab, logits + size_t(i)*n_vocab, base_log_probs.data() + size_t(i)*nv, tokens[i+1], local_kld);
             kld_values[i]    = (float)v.first;
             p_diff_values[i] = v.second;
         }
```

Let me know if there's a case where it still crashes.

**Comment by feffy380** [comment]
So far not seeing any segfaults with the patch 👍
Tested 32k and 64k context
